In [1]:
import numpy as np
import pandas as pd

import viz # curriculum example visualizations

np.random.seed(21)

### 1. How likely is it that you roll doubles when rolling two dice?

In [31]:
n1_trials = n1_rows = 10**5
n1_dice = n1_cols = 2

rolls1 = np.random.choice([1,2,3,4,5,6], n1_trials * n1_dice).reshape(n1_rows,n1_cols)

doubles_count = pd.DataFrame(rolls1).apply(lambda row: row[0] == row[1], axis = 1).sum()

prob_of_doubles = doubles_count / n1_rows
prob_of_doubles

0.16668

ANS: The probability is 16.67% (1 in 6)

### 2. If you flip 8 coins, what is the probability of getting exactly 3 heads? What is the probability of getting more than 3 heads?

In [41]:
n2_trials = n2_rows = 10**5
n2_coins = n2_cols = 8

#Will set up the binary options as heads = 1, tails = 0
flips2 = np.random.choice([0,1],n2_trials * n2_coins).reshape(n2_rows,n2_cols)

sums_of_flips = flips2.sum(axis=1)

exactly_3_heads = (sums_of_flips == 3).sum()
exactly_3_heads/n2_rows

0.21913

You will get exactly 3 heads 21.9% of the time.

In [40]:
more_than_3_heads = (sums_of_flips > 3).sum() 
more_than_3_heads/n2_rows

0.63598

You will get more than 3 heads 63.6% of the time.

### 3. There are approximitely 3 web development cohorts for every 1 data science cohort at Codeup. Assuming that Codeup randomly selects an alumni to put on a billboard, what are the odds that the two billboards I drive past both have data science students on them?

In [55]:
n3_trials = n3_rows = 10**5
n3_boards = n3_cols = 2

#Assuming all cohorts are the same size:
p_ds = .25

n3_data = np.random.random((n3_rows, n3_cols))
n3_results = ((n3_data < p_ds).sum(axis = 1)) == 2
n3_results.sum()/n3_rows

0.06235

The odds of passing two data science alum billboards is 6.24%.

### 4. Codeup students buy, on average, 3 poptart packages with a standard deviation of 1.5 a day from the snack vending machine. If on monday the machine is restocked with 17 poptart packages, how likely is it that I will be able to buy some poptarts on Friday afternoon? (Remember, if you have mean and standard deviation, use the np.random.normal)

In [61]:
n4_trials = n4_rows = 10**5
# columns = 5, for 5 days from Monday to Friday.
#Assume they load first thing Monday morning, the draw is the last draw of the week before restock the following Monday, and that it is always stocked to 17.

poptarts = np.random.normal(3,1.5,(n4_rows,5))
still_tartin = (poptarts.sum(axis=1) < 17).sum()
still_tartin/n4_rows


0.7217

You will be able to pop that tart on 72.2% of Friday afternoons.

### 5. Men have an average height of 178 cm and standard deviation of 8cm. Women have a mean of 170, sd = 6cm. Since you have means and standard deviations, you can use np.random.normal to generate observations.  If a man and woman are chosen at random, what is the likelihood the woman is taller than the man?

In [77]:
n5_trials = n5_rows = 10**5

tallness = np.random.normal([178,170],[8,6],(n5_rows,2))

women_taller = pd.DataFrame(tallness).apply(lambda row: row[0] < row[1], axis = 1).sum()
women_taller/n5_rows

0.21146

They are likely taller 21% of the time.

### 6. When installing anaconda on a student's computer, there's a 1 in 250 chance that the download is corrupted and the installation fails. 

In [80]:
curruption = 1/250

#### (6a) What are the odds that after having 50 students download anaconda, no one has an installation issue? 100 students?

In [98]:
(np.random.random(50) < curruption).sum()

0

In [104]:
(np.random.random(100) < curruption).sum()

0

In [105]:
(np.random.random(100000) < curruption).sum()

405

I ran this many times.  50 students never showed curruption, and every 8 or 10 runs of the 100 would pop a single curruption.  Running it with 100k to get a decent probability came to 405, or 0.4%.

#### (6b) What is the probability that we observe an installation issue within the first 150 students that download anaconda?

In [113]:
random_150 = (np.random.random(150) < curruption).sum()

0.0

In [143]:
# wasn't satisfied so I simulated the simulation 100k times to see how many produced a curruption:
thing = []
for i in range(100000):
    if ((np.random.random(150) < curruption).sum()) > 0:
        thing.append(1)
result7b = len(thing)/100000
result7b


0.44883

#### (6c) How likely is it that 450 students all download anaconda without an issue?

In [148]:
random_450 = (np.random.random(450) < curruption).sum()
random_450

thing2 = []
for i in range(100000):
    if ((np.random.random(450) < curruption).sum()) > 0:
        thing2.append(1)
1-(len(thing)/100000)

0.5511699999999999

55.1%

### 7. There's a 70% chance on any given day that there will be at least one food truck at Travis Park. However, you haven't seen a food truck there in 3 days. How unlikely is this?

In [162]:
n7_trials = n7_rows = 10**5
n7_days = n7_cols = 3
p_foodtruck = .7

((np.random.random((n7_rows,n7_cols)) > p_foodtruck).sum(axis=1) == 3).sum()/n7_rows

0.02662

2.67% chance it's not there three days in a row.
#### (7b) How likely is it that a food truck will show up sometime this week?

In [164]:
n7_days_b = n7_cols_b = 7

((np.random.random((n7_rows,n7_cols_b)) < p_foodtruck).sum(axis=1) > 1).sum()/n7_rows

0.99625

99.6% chance it will show up

### 8. If 23 people are in the same room, what are the odds that two of them share a birthday? What if it's 20 people? 40?

In [168]:
n8_rooms = n8_rows = 10**5
n8_room = n8_cols = 23
p_birthday = 1/365

((np.random.random((n8_rows,n8_cols)) < p_birthday).sum(axis = 1) > 1).sum()/n8_rows

0.00193

In [169]:
n8_room = 20
((np.random.random((n8_rows,n8_cols)) < p_birthday).sum(axis = 1) > 1).sum()/n8_rows

0.00184

In [170]:
n8_room = 40
((np.random.random((n8_rows,n8_cols)) < p_birthday).sum(axis = 1) > 1).sum()/n8_rows

0.00207

23 people = 1.93%, 20 people = 1.84%, 40 people = 2.07%